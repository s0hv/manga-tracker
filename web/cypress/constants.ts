export const nextAuthSessionCookie = 'next-auth.session-token';
