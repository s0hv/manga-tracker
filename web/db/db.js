const { db } = require('.');

function getLatestReleases(serviceId, mangaId, userUUID) {
  const joins = [];
  const where = [];
  const args = [];
  let paramNumber = 1;
  if (userUUID) {
    joins.push(`INNER JOIN user_follows uf ON c.manga_id = uf.manga_id AND (uf.service_id IS NULL OR c.service_id=uf.service_id) 
                    INNER JOIN users u ON u.user_id=uf.user_id`);
    where.push(`u.user_uuid=$${paramNumber}::uuid`);
    args.push(userUUID);
    paramNumber++;
  }

  if (mangaId) {
    where.push(`c.manga_id=$${paramNumber}`);
    args.push(mangaId);
    paramNumber++;
  }

  if (serviceId) {
    where.push(`c.service_id=$${paramNumber}`);
    args.push(serviceId);
    paramNumber++;
  }

  const sql = `
        WITH chapters_filtered AS (
            SELECT chapter_id, title, chapter_number, chapter_decimal, release_date, chapter_identifier, c.service_id, c.manga_id, g.name as "group"
            FROM chapters as c
            INNER JOIN groups g ON g.group_id = c.group_id 
            ${joins.join(' ')}
            ${where.length > 0 ? 'WHERE ' + where.join(' AND ') : ''}
        )
        SELECT 
               c.chapter_id,
               m.title as manga_title,
               m.manga_id,
               m.release_interval,
               c.title,
               c.chapter_number,
               c.chapter_decimal,
               c.release_date,
               c.chapter_identifier,
               s.service_name,
               s.chapter_url_format,
               s.url,
               c."group",
               mi.cover
        FROM chapters_filtered c 
            INNER JOIN manga m on c.manga_id = m.manga_id
            INNER JOIN services s on c.service_id = s.service_id
            LEFT JOIN manga_info mi ON m.manga_id = mi.manga_id
        WHERE c.release_date > NOW() - INTERVAL '1 hour'
        UNION 
              (SELECT 
                      c.chapter_id,
                      m.title as manga_title,
                      m.manga_id,
                      m.release_interval,
                      c.title,
                      c.chapter_number,
                      c.chapter_decimal,
                      c.release_date,
                      c.chapter_identifier,
                      s.service_name,
                      s.chapter_url_format,
                      s.url,
                      c."group",
                      mi.cover
              FROM chapters_filtered c
                  INNER JOIN manga m on c.manga_id = m.manga_id
                  INNER JOIN services s on c.service_id = s.service_id
                  LEFT JOIN manga_info mi ON m.manga_id = mi.manga_id
              ORDER BY release_date DESC, chapter_number DESC
              LIMIT 30)
        ORDER BY release_date DESC, chapter_number DESC`;

  return db.any(sql, args);
}

function getUserFollows(userId, mangaId) {
  const sql = 'SELECT service_id FROM user_follows WHERE user_id=$1 AND manga_id=$2';
  return db.query(sql, [userId, mangaId]);
}

module.exports = { getLatestReleases, getUserFollows };
