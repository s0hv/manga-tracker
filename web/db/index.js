const camelcaseKeys = require('camelcase-keys');
const { performance } = require('perf_hooks');

const { queryLogger } = require('../utils/logging');

const isTest = process.env.NODE_ENV === 'test';

// eslint-disable-next-line import/order
const pgp = require('pg-promise')({
  noLocking: isTest, // Locking must not be set on during test so mocks can be made.
  capSQL: true,
  pgFormatting: true, // When this is false parameters would be always logged since they're embedded in the query

  // Log duration and query
  receive(data, result, e) {
    const { duration } = result;

    // This should result in minimal performance impact except on the first query,
    // as the object keys are not yet cached at that point
    const t0 = performance.now();
    result.rows = camelcaseKeys(data, { deep: true });
    const camelCaseDuration = performance.now() - t0;

    if (queryLogger.level === 'debug') {
      queryLogger.debug({
        params: e.params,
        duration: `${duration}ms`,
        camelCaseDuration,
      }, e.query);
    } else {
      queryLogger.info({
        duration: `${duration}ms`,
        camelCaseDuration,
      }, e.query);
    }
  },

  // Error logging
  error(err, e) {
    queryLogger.info(e, 'Timeout debug');
    queryLogger.error(err, e.query);
  },
});

// https://stackoverflow.com/a/34427278/6046713
const createSingletonDb = () => {
  const s = Symbol.for('database');
  let scope = global[s];
  if (!global[s]) {
    scope = pgp({
      host: process.env.DB_HOST,
      user: process.env.DB_USER,
      database: isTest ?
        process.env.DB_NAME_TEST || process.env.DB_NAME :
        process.env.DB_NAME,
      port: process.env.DB_PORT,
      password: process.env.PGPASSWORD,
      max: 5,
      idleTimeoutMillis: 30000,
      connectionTimeoutMillis: 5000,
    });
    global[s] = scope;
  }
  return scope;
};

/**
 * Database object
 * @type {pgPromise.IDatabase}
 */
const db = createSingletonDb();


function query(sql, args) {
  return db.query(sql, args);
}

async function end() {
  await pgp.end();
}

module.exports = {
  query,
  db,
  pgp,
  end,
};
