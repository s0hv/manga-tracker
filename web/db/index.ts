import camelcaseKeys from 'camelcase-keys';
import { performance } from 'perf_hooks';
import pgPromise, { IDatabase, QueryParam } from 'pg-promise';

import { queryLogger } from '../utils/logging.js';

const isTest = process.env.NODE_ENV === 'test';

export const pgp = pgPromise({
  noLocking: isTest, // Locking must not be set on during test so mocks can be made.
  capSQL: true,
  pgFormatting: true, // When this is false parameters would always be logged since they're embedded in the query

  // Log duration and query
  receive(data, result, e) {
    if (!result) return;

    const { duration } = result;

    // This should result in minimal performance impact except on the first query,
    // as the object keys are not yet cached at that point
    const t0 = performance.now();
    result.rows = camelcaseKeys(data, { deep: true });
    const camelCaseDuration = performance.now() - t0;

    if (queryLogger.level === 'debug') {
      queryLogger.debug({
        params: e.params,
        duration: `${duration}ms`,
        camelCaseDuration,
      }, e.query);
    } else {
      queryLogger.info({
        duration: `${duration}ms`,
        camelCaseDuration,
      }, e.query);
    }
  },

  // Error logging
  error(err, e) {
    queryLogger.error(err, e.query);
  },
});

// https://stackoverflow.com/a/34427278/6046713
const createSingletonDb = (): IDatabase<any> => {
  const s = Symbol.for('database');
  let scope = global[s];
  if (!global[s]) {
    scope = pgp({
      host: process.env.DB_HOST,
      user: process.env.DB_USER,
      database: isTest ?
        process.env.DB_NAME_TEST || process.env.DB_NAME :
        process.env.DB_NAME,
      port: Number(process.env.DB_PORT),
      password: process.env.PGPASSWORD,
      max: 5,
      idleTimeoutMillis: 30000,
      connectionTimeoutMillis: 5000,
    });
    global[s] = scope;
  }
  return scope;
};

export const db = createSingletonDb();


export function query(sql: QueryParam, args?: any) {
  return db.query(sql, args);
}

export async function end() {
  await pgp.end();
}
