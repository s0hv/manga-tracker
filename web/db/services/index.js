import snakecaseKeys from 'snakecase-keys';
import { db } from '..';
import { generateUpdate } from '../utils.js';


export const getService = (serviceId) => db.oneOrNone('SELECT * FROM services WHERE service_id=$1', [serviceId]);

export const getServiceWhole = (serviceId) => db.oneOrNone('SELECT * FROM service_whole WHERE service_id=$1', [serviceId]);

export const getServiceConfig = (serviceId) => db.oneOrNone('SELECT * FROM service_config WHERE service_id=$1', [serviceId]);

export const getServiceFull = (serviceId) => {
  const retVal = {};

  return Promise.all([
    getService(serviceId).then(row => { retVal.service = row }),
    getServiceWhole(serviceId).then(row => { retVal.serviceWhole = row }),
    getServiceConfig(serviceId).then(row => { retVal.serviceConfig = row }),
  ])
    .then(() => retVal);
};

/**
 * Update service row
 * @returns {Promise<pgPromise.IResultExt>}
 */
export const updateService = ({
  serviceId,
  serviceName,
  url,
  chapterUrlFormat,
  mangaUrlFormat,
  disabled,
  disabledUntil,
}) => {
  const service = {
    service_name: serviceName,
    url,
    chapter_url_format: chapterUrlFormat,
    manga_url_format: mangaUrlFormat,
    disabled,
    disabled_until: disabledUntil,
  };

  const sql = `${generateUpdate(service, 'services')} WHERE service_id=$1`;
  return db.result(sql, [serviceId]);
};

/**
 * Update service_whole row
 * @returns {Promise<pgPromise.IResultExt>}
 */
export const updateServiceWhole = ({
  serviceId,
  feedUrl,
  nextUpdate,
  lastId,
}) => {
  const serviceWhole = {
    feed_url: feedUrl,
    next_update: nextUpdate,
    last_id: lastId,
  };

  const sql = `${generateUpdate(serviceWhole, 'service_whole')} WHERE service_id=$1`;
  return db.result(sql, [serviceId]);
};

/**
 * Update service_config row
 * @returns {Promise<pgPromise.IResultExt>}
 */
export const updateServiceConfig = ({
  serviceId,
  checkInterval,
  scheduledRunInterval,
  scheduledRunLimit,
  scheduledRunsEnabled,
}) => {
  const serviceConfig = snakecaseKeys({
    checkInterval,
    scheduledRunInterval,
    scheduledRunLimit,
    scheduledRunsEnabled,
  });
  const sql = `${generateUpdate(serviceConfig, 'service_config')} WHERE service_id=$1`;
  return db.result(sql, [serviceId]);
};

/**
 * Get all service configs
 * @returns {Promise<any[]>}
 */
export const getServiceConfigs = () => {
  const sql = 'SELECT * FROM service_config';
  return db.many(sql);
};

