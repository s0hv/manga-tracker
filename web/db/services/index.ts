import { db } from '../helpers';
import { generateUpdate } from '../utils';

import type { DatabaseId } from '@/types/dbTypes';
import type { Service, ServiceConfig, ServiceWhole } from '@/types/db/services';
import type { PartialExcept } from '@/types/utility';


export const getService = (serviceId: DatabaseId) => db.oneOrNone<Service>`SELECT * FROM services WHERE service_id=${serviceId}`;

export const getServiceWhole = (serviceId: DatabaseId) => db.oneOrNone<ServiceWhole>`SELECT * FROM service_whole WHERE service_id=${serviceId}`;

export const getServiceConfig = (serviceId: DatabaseId) => db.oneOrNone<ServiceConfig>`SELECT * FROM service_config WHERE service_id=${serviceId}`;

export const getServiceFull = (serviceId: DatabaseId) => {
  const retVal: {
    service?: Service | null,
    serviceWhole?: ServiceWhole | null,
    serviceConfig?: ServiceConfig | null
  } = {};

  return Promise.all([
    getService(serviceId).then(row => { retVal.service = row }),
    getServiceWhole(serviceId).then(row => { retVal.serviceWhole = row }),
    getServiceConfig(serviceId).then(row => { retVal.serviceConfig = row }),
  ])
    .then(() => retVal);
};

/**
 * Update service row
 */
export const updateService = ({
  serviceId,
  serviceName,
  url,
  chapterUrlFormat,
  mangaUrlFormat,
  disabled,
  disabledUntil,
}: PartialExcept<Service, 'serviceId'>) => {
  const service = {
    service_name: serviceName,
    url,
    chapter_url_format: chapterUrlFormat,
    manga_url_format: mangaUrlFormat,
    disabled,
    disabled_until: disabledUntil,
  };

  return db.sql`UPDATE services SET ${generateUpdate(service, db.sql)} WHERE service_id=${serviceId}`
    .execute();
};

/**
 * Update service_whole row
 */
export const updateServiceWhole = ({
  serviceId,
  feedUrl,
  nextUpdate,
  lastId,
}: PartialExcept<ServiceWhole, 'serviceId'>) => {
  const serviceWhole = {
    feed_url: feedUrl,
    next_update: nextUpdate,
    last_id: lastId,
  };

  return db.sql`UPDATE service_whole SET ${generateUpdate(serviceWhole, db.sql)} WHERE service_id=${serviceId}`
    .execute();
};

/**
 * Update service_config row
 */
export const updateServiceConfig = ({
  serviceId,
  checkInterval,
  scheduledRunInterval,
  scheduledRunLimit,
  scheduledRunsEnabled,
}: PartialExcept<ServiceConfig, 'serviceId'>) => {
  const serviceConfig = {
    checkInterval,
    scheduledRunInterval,
    scheduledRunLimit,
    scheduledRunsEnabled,
  };
  return db.sql`UPDATE service_config SET ${generateUpdate(serviceConfig, db.sql)} WHERE service_id=${serviceId}`
    .execute();
};

/**
 * Get all service configs
 */
export const getServiceConfigs = (): Promise<ServiceConfig[]> => {
  return db.many<ServiceConfig>`SELECT * FROM service_config`;
};

