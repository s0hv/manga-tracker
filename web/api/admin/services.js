import { body, matchedData, param } from 'express-validator';

import { requiresUser } from '@/db/auth';
import { handleError } from '@/db/utils';
import {
  updateService,
  updateServiceWhole,
  updateServiceConfig,
} from '@/db/services';
import {
  validateAdminUser,
  handleValidationErrors,
  serviceIdValidation,
  isISO8601Duration,
} from '../../utils/validators.js';

export default app => {
  const validateService = [
    body('service')
      .isObject()
      .optional(),
    body('service.serviceName')
      .isString()
      .optional(),
    body('service.disabled')
      .isBoolean({ strict: true })
      .optional()
      .toBoolean(true),

    body('serviceWhole')
      .isObject()
      .optional(),
    body('serviceWhole.nextUpdate')
      .isISO8601({ strict: true })
      .optional({ nullable: true })
      .toDate(),

    body('serviceConfig')
      .isObject()
      .optional(),
    isISO8601Duration(body('serviceConfig.checkInterval'))
      .optional(),
    body('serviceConfig.scheduledRunLimit')
      .isInt({ min: 1, max: 100 })
      .optional()
      .withMessage('scheduledRunLimit must be between 1 and 100'),
    body('serviceConfig.scheduledRunsEnabled')
      .isBoolean({ strict: true })
      .optional()
      .toBoolean(),
    isISO8601Duration(body('serviceConfig.scheduledRunInterval'))
      .optional(),
  ];

  app.post('/api/admin/editService/:serviceId', requiresUser, [
    validateAdminUser(),
    serviceIdValidation(param('serviceId')),
    ...validateService,
    handleValidationErrors,
  ], (req, res) => {
    const {
      service,
      serviceWhole,
      serviceConfig,
    } = matchedData(req, { includeOptionals: false, onlyValidData: true });

    const serviceId = req.params.serviceId;

    const serviceArgsExist = service && Object.keys(service).length !== 0;
    const serviceWholeArgsExist = serviceWhole && Object.keys(serviceWhole).length !== 0;
    const serviceConfigArgsExist = serviceConfig && Object.keys(serviceConfig).length !== 0;

    if (!serviceArgsExist && !serviceWholeArgsExist && !serviceConfigArgsExist) {
      res.status(400).json({ error: 'No valid fields given to update' });
      return;
    }

    req.log.info('Updating service with %o', req.body);

    const promise = Promise.resolve();

    if (serviceWholeArgsExist) {
      promise.then(() => updateServiceWhole({ ...serviceWhole, serviceId }));
    }

    if (serviceArgsExist) {
      promise.then(() => updateService({ ...service, serviceId }));
    }

    if (serviceConfigArgsExist) {
      promise.then(() => updateServiceConfig({ ...serviceConfig, serviceId }));
    }

    promise
      .then(() => res.json({ message: 'OK' }))
      .catch(err => handleError(err, res));
  });
};
