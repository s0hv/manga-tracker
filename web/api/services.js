import { handleError } from '@/db/utils';
import { getServicesForApi } from '@/db/services/serviceInfo';


export default app => {
  /**
   *  @openapi
   *  /services:
   *    get:
   *      summary: Get all services that might have chapters associated with them
   *      responses:
   *        200:
   *          description: Returns the list of services
   *          content:
   *            application/json:
   *              schema:
   *                type: object
   *                properties:
   *                  data:
   *                    type: array
   *                    items:
   *                      $ref: '#/components/schemas/service'
   *                required:
   *                  - data
   */
  app.get('/api/services', (req, res) => {
    getServicesForApi()
      .then(services => res.json({ data: services }))
      .catch(err => handleError(err, res));
  });
};
