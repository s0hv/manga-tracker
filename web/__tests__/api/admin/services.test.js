import request from 'supertest';
import { parse, toSeconds } from 'iso8601-duration';

import { csrfMissing } from '../../../utils/constants';

import { userForbidden, userUnauthorized, isCI } from '../../constants';
import initServer from '../../initServer';
import stopServer from '../../stopServer';
import {
  adminUser,
  normalUser,
  withUser,
  expectErrorMessage,
  getErrorMessage,
  getIncrementalStringGenerator,
  mockUTCDates,
} from '../../utils';
import { getServiceFull } from '../../../db/services';
import { copyService } from '../../dbutils';

let httpServer;

beforeAll(async () => {
  ({ httpServer } = await initServer());
});

afterAll(async () => stopServer(httpServer));

describe('POST /api/admin/editService/:serviceId', () => {
  mockUTCDates();
  const url = '/api/admin/editService/3';
  const serviceId = 3;
  const baseBody = {
    service: {
      serviceName: undefined,
      disabled: undefined,
    },
    serviceWhole: {
      nextUpdate: undefined,
    },
    serviceConfig: {
      checkInterval: undefined,
      scheduledRunLimit: undefined,
      scheduledRunsEnabled: undefined,
      scheduledRunInterval: undefined,
    },
  };

  it('Returns 403 without CSRF token', async () => {
    await request(httpServer)
      .post(url)
      .expect(403)
      .expect(expectErrorMessage(csrfMissing));
  });

  it('returns unauthorized without user', async () => {
    await request(httpServer)
      .post(url)
      .csrf()
      .expect(401)
      .expect(expectErrorMessage(userUnauthorized));
  });

  it('returns forbidden for non admin', async () => {
    await withUser(normalUser, async () => {
      await request(httpServer)
        .post(url)
        .csrf()
        .expect(403)
        .expect(expectErrorMessage(userForbidden));
    });
  });

  it('returns 400 for admin without data', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post(url)
        .csrf()
        .expect(400)
        .expect(expectErrorMessage('No valid fields given to update'));

      await request(httpServer)
        .post(url)
        .csrf()
        .send({})
        .expect(400)
        .expect(expectErrorMessage('No valid fields given to update'));

      await request(httpServer)
        .post(url)
        .csrf()
        .send(baseBody)
        .expect(400)
        .expect(expectErrorMessage('No valid fields given to update'));
    });
  });

  it('Returns 400 with invalid service data', async () => {
    await withUser(adminUser, async () => {
      // service_name
      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          service: {
            serviceName: [1, 2, 3],
          },
        })
        .expect(400)
        .expect(expectErrorMessage([1, 2, 3], 'service.serviceName'));

      // disabled
      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          service: {
            disabled: '2',
          },
        })
        .expect(400)
        .expect(expectErrorMessage('2', 'service.disabled'));

      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          service: {
            disabled: null,
          },
        })
        .expect(400)
        .expect(expectErrorMessage(null, 'service.disabled'));
    });
  });

  it('Returns 400 with invalid service whole data', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          serviceWhole: {
            nextUpdate: 'abc',
          },
        })
        .expect(400)
        .expect(expectErrorMessage('abc', 'serviceWhole.nextUpdate'));

      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          serviceWhole: {
            nextUpdate: 1602954767,
          },
        })
        .expect(400)
        .expect(expectErrorMessage(1602954767, 'serviceWhole.nextUpdate'));
    });
  });

  it('Returns 400 with invalid service config data', async () => {
    await withUser(adminUser, async () => {
      // checkInterval and scheduledRunInterval
      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          serviceConfig: {
            checkInterval: '3Y6M4DT12H30M5S',
            scheduledRunInterval: '3Y6M4DT12H30M5S',
          },
        })
        .expect(400)
        .expect(res => expect(getErrorMessage(res)).toMatchSnapshot());

      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          serviceConfig: {
            checkInterval: 'T12H30M5S',
            scheduledRunInterval: 'T12H30M5S',
          },
        })
        .expect(400)
        .expect(res => expect(getErrorMessage(res)).toMatchSnapshot());


      // scheduledRunLimit
      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          serviceConfig: {
            scheduledRunLimit: 'a',
          },
        })
        .expect(400)
        .expect(res => expect(getErrorMessage(res)).toMatchSnapshot());

      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          serviceConfig: {
            scheduledRunLimit: 101,
          },
        })
        .expect(400)
        .expect(res => expect(getErrorMessage(res)).toMatchSnapshot());

      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          serviceConfig: {
            scheduledRunLimit: 0,
          },
        })
        .expect(400)
        .expect(res => expect(getErrorMessage(res)).toMatchSnapshot());


      // scheduledRunsEnabled
      await request(httpServer)
        .post(url)
        .csrf()
        .send({
          serviceConfig: {
            scheduledRunsEnabled: 'y',
          },
        })
        .expect(400)
        .expect(expectErrorMessage('y', 'serviceConfig.scheduledRunsEnabled'));
    });
  });

  const expectUpdateDoneCorrectly = async (data) => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post(url)
        .csrf()
        .send(data)
        .expect(200);

      await new Promise(setImmediate);
      // Sleep for a bit here to make sure changes get flushed to database
      await new Promise(r => setTimeout(r, 20));

      const newServiceId = isCI ? await copyService(serviceId) : serviceId;

      const originalService = await getServiceFull(newServiceId);
      originalService.serviceConfig.checkInterval = toSeconds(originalService.serviceConfig.checkInterval);
      originalService.serviceConfig.scheduledRunInterval = toSeconds(originalService.serviceConfig.scheduledRunInterval);

      if (data.service) {
        expect(originalService.service).toMatchObject(data.service);
      }
      if (data.serviceWhole) {
        expect(originalService.serviceWhole).toMatchObject(data.serviceWhole);
      }
      if (data.serviceConfig) {
        expect(originalService.serviceConfig).toMatchObject({
          ...data.serviceConfig,
          checkInterval: toSeconds(parse(data.serviceConfig.checkInterval)),
          scheduledRunInterval: toSeconds(parse(data.serviceConfig.scheduledRunInterval)),
        });
      }
    });
  };
  const nameGen = getIncrementalStringGenerator('servicesApi');

  it('returns ok when editing with valid data full', async () => {
    await expectUpdateDoneCorrectly({
      service: {
        serviceName: nameGen(),
        disabled: true,
      },
      serviceWhole: {
        nextUpdate: new Date(1602954767000),
      },
      serviceConfig: {
        checkInterval: 'P1Y1M1DT1H1M1S',
        scheduledRunInterval: 'P0Y6M1DT0H30M5S',
        scheduledRunLimit: 5,
        scheduledRunsEnabled: false,
      },
    });
  });

  it('returns ok when editing with only service data', async () => {
    await expectUpdateDoneCorrectly(
      {
        service: {
          serviceName: nameGen(),
          disabled: false,
        },
      }
    );
  });

  it('returns ok when editing only service whole data', async () => {
    await expectUpdateDoneCorrectly({
      serviceWhole: {
        nextUpdate: new Date(1702954767000),
      },
    });
  });

  it('returns ok when editing only service config data', async () => {
    await expectUpdateDoneCorrectly({
      serviceConfig: {
        checkInterval: 'P1Y0M1DT1H0M5S',
        scheduledRunInterval: 'P0Y0M4DT12H3M8S',
        scheduledRunLimit: 10,
        scheduledRunsEnabled: true,
      },
    });
  });
});
