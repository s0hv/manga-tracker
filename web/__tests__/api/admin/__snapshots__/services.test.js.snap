// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`POST /api/admin/editService/:serviceId > Returns 400 with invalid service config data 1`] = `
[
  {
    "location": "body",
    "msg": "Value must be a valid ISO 8601 duration",
    "param": "serviceConfig.checkInterval",
    "value": "3Y6M4DT12H30M5S",
  },
  {
    "location": "body",
    "msg": "Value must be a valid ISO 8601 duration",
    "param": "serviceConfig.scheduledRunInterval",
    "value": "3Y6M4DT12H30M5S",
  },
]
`;

exports[`POST /api/admin/editService/:serviceId > Returns 400 with invalid service config data 2`] = `
[
  {
    "location": "body",
    "msg": "Value must be a valid ISO 8601 duration",
    "param": "serviceConfig.checkInterval",
    "value": "T12H30M5S",
  },
  {
    "location": "body",
    "msg": "Value must be a valid ISO 8601 duration",
    "param": "serviceConfig.scheduledRunInterval",
    "value": "T12H30M5S",
  },
]
`;

exports[`POST /api/admin/editService/:serviceId > Returns 400 with invalid service config data 3`] = `
[
  {
    "location": "body",
    "msg": "scheduledRunLimit must be between 1 and 100",
    "param": "serviceConfig.scheduledRunLimit",
    "value": "a",
  },
]
`;

exports[`POST /api/admin/editService/:serviceId > Returns 400 with invalid service config data 4`] = `
[
  {
    "location": "body",
    "msg": "scheduledRunLimit must be between 1 and 100",
    "param": "serviceConfig.scheduledRunLimit",
    "value": 101,
  },
]
`;

exports[`POST /api/admin/editService/:serviceId > Returns 400 with invalid service config data 5`] = `
[
  {
    "location": "body",
    "msg": "scheduledRunLimit must be between 1 and 100",
    "param": "serviceConfig.scheduledRunLimit",
    "value": 0,
  },
]
`;
